src/upload.ts

```diff
@@ -0,0 +1,150 @@
+import {
+  Instrumentation, IPayload, ITotalByteLimitExceededEventState, IUploadInfo, IXhrErrorEventState,
+  State, UploadCallback
+} from "../clarity";
+import { config } from "./config";
+import { instrument, state, teardown } from "./core";
+import { debug, getCookie, getEventId, guid, isNumber, mapProperties, setCookie } from "./utils";
+
+// Counters
+let sentBytesCount: number;
+let uploadCount: number;
+
+// Storage for payloads that are compressed and are ready to be sent, but are waiting for Clarity trigger.
+// Once trigger is fired, all payloads from this queue will be sent in the order they were generated.
+let uploadQueue: IUploadInfo[];
+
+// Map from a sequence number to an UploadInfo object for payloads that were attempted to be sent, but were not succesfully delivered yet
+let uploadInfos: { [key: number]: IUploadInfo };
+
+let reUploadQueue: number[];
+
+export function upload(compressed: string, raw: IPayload, onSuccessExtra?: UploadCallback, onFailureExtra?: UploadCallback) {
+  let uploadHandler = config.uploadHandler || defaultUpload;
+  let sequenceNo = raw.envelope.sequenceNumber;
+  let onSuccess = (status: number) => {
+    onSuccessDefault(sequenceNo, status);
+    if (onSuccessExtra) {
+      onSuccessExtra(status);
+    }
+  };
+  let onFailure = (status: number) => {
+    onFailureDefault(sequenceNo, status);
+    if (onFailureExtra) {
+      onFailureExtra(status);
+    }
+  };
+  uploadInfos[sequenceNo] = { compressed, raw, failureCount: 0 };
+  uploadHandler(compressed, onSuccess, onFailure);
+
+  if (config.debug) {
+    let rawLength = JSON.stringify(raw).length;
+    debug(`** Upload #${uploadCount}: Batch #${sequenceNo}`
+          + ` ${Math.round(compressed.length / 1024.0)}KB`
+          + ` (${Math.round(rawLength / 1024.0)}KB Raw). **`);
+  }
+  uploadCount++;
+}
+
+export function enqueueUpload(compressed: string, raw: IPayload) {
+  let payloadInfo: IUploadInfo = { compressed, raw, failureCount: 0 };
+  uploadQueue.push(payloadInfo);
+}
+
+export function flushUploadQueue() {
+  uploadSequentially(uploadQueue);
+}
+
+export function uploadSequentially(uploads: IUploadInfo[]) {
+  if (uploads.length > 0) {
+    let nextUpload = uploads.shift();
+    let uploadNext = (status: number) => {
+      uploadSequentially(uploads);
+    };
+    upload(nextUpload.compressed, nextUpload.raw, uploadNext, uploadNext);
```

Comment:
>upload(nextUpload.compressed, nextUpload.raw, uploadNext, uploadNext); [](start = 4, length = 70)

the naming here makes it difficult to parse - how about keeping upload as a verb and using payload (or some other noun) to disambiguate. As it stands now you check an array of uploads then call upload on the nextUpload with an uploadNext callback, way too much of the same word being used to mean different things

Reply:
Renamed

---
In reply to: [168001870](https://github.com/Microsoft/clarity-js/pull/63#discussion_r168001870) [](ancestors = 168001870)